runtime/router.go

```diff
@@ -133,47 +120,60 @@ func NewRouterEndpoint(
 	endpointName string,
 	handlerName string,
 	handler HandlerFn,
+	scope tally.Scope,
 ) *RouterEndpoint {
-	endpointTags := map[string]string{
-		"endpoint": endpointName,
-		"handler":  handlerName,
-	}
-	endpointScope := gateway.AllHostScope.Tagged(endpointTags)
-	requestRecvd := endpointScope.Counter("inbound.calls.recvd")
-	requestLatency := endpointScope.Timer("inbound.calls.latency")
-	statusCodes := make(map[int]tally.Counter, len(knownStatusCodes))
-
-	for _, statusCode := range knownStatusCodes {
-		metricName := "inbound.calls.status." + strconv.Itoa(statusCode)
-		statusCodes[statusCode] = endpointScope.Counter(metricName)
-	}
-
 	return &RouterEndpoint{
 		EndpointName: endpointName,
 		HandlerName:  handlerName,
 		HandlerFn:    handler,
+		scope:        scope,
 		gateway:      gateway,
-
-		metrics: EndpointMetrics{
-			requestRecvd:   requestRecvd,
-			statusCodes:    statusCodes,
-			requestLatency: requestLatency,
-		},
 	}
 }
 
 // HandleRequest is called by the router and starts the request
 func (endpoint *RouterEndpoint) HandleRequest(
-	w http.ResponseWriter, r *http.Request, params httprouter.Params,
+	w http.ResponseWriter,
+	r *http.Request,
+	params httprouter.Params,
 ) {
-	reqFields := logRequestFields(r)
+	startTime := time.Now()
+	var req *ServerHTTPRequest
 	var resFields []zapcore.Field
+	reqFields := logRequestFields(r)
 
 	defer func() {
-		writeLogs(endpoint.gateway.Logger, reqFields, resFields)
+		writeLogs(req.Logger, reqFields, resFields)
+
+		if _, ok := knownStatusCodes[req.res.StatusCode]; !ok {
+			req.Logger.Error(
+				"Unexpected status code",
+				zap.Int("UnexpectedStatusCode", req.res.StatusCode),
+			)
+		}
+
+		scope := endpoint.scope.Tagged(map[string]string{
+			"status-code": strconv.Itoa(req.res.StatusCode),
+			//"calling-service": "unknown", // TODO: Add 'calling-service' tag to scope
+		})
+		scope.Counter("inbound.calls.recvd").Inc(1)
+		if req.res.StatusCode >= 400 && req.res.StatusCode < 600 {
+			scope.Counter("inbound.calls.errors").Inc(1)
+			if req.res.StatusCode >= 400 && req.res.StatusCode < 500 {
+				scope.Counter("inbound.calls.status.4XX").Inc(1)
```

Comment:
nit: assign const vars for strings like "inbound.calls.status.4XX"

Reply:
Done.